src/advertise.rs

```diff
@@ -0,0 +1,271 @@
+/// Call [ServiceBuilder::finalise] to publish it.
+pub struct ServiceBuilder<'a> {
+    mdns: &'a ServiceDaemon,
+    ips: Vec<Ipv4Addr>,
+    hostname: String,
+    ty: ThingType,
+    port: u16,
+    path: String,
+    name: String,
+}
+
+impl<'a> ServiceBuilder<'a> {
+    fn new(ad: &'a Advertiser, name: impl AsRef<str>) -> ServiceBuilder<'a> {
+        Self {
+            name: name.as_ref().to_string(),
+            mdns: &ad.mdns,
+            ips: ad.ips.clone(),
+            hostname: ad.hostname.clone(),
+            ty: ThingType::Thing,
+            port: 8080,
+            path: WELL_KNOWN.to_string(),
+        }
+    }
+
+    pub fn thing_type(mut self, ty: ThingType) -> Self {
+        self.ty = ty;
+
+        self
+    }
+
+    pub fn port(mut self, port: u16) -> Self {
+        self.port = port;
+
+        self
+    }
+
+    pub fn path(mut self, path: impl AsRef<str>) -> Self {
+        self.path = path.as_ref().to_string();
+
+        self
+    }
+
+    pub fn hostname(mut self, host: impl AsRef<str>) -> Self {
+        self.hostname = host.as_ref().to_string();
+
+        self
+    }
+
+    pub fn ips<I: Into<Ipv4Addr>>(mut self, ips: impl Iterator<Item = I>) -> Self {
+        self.ips = ips.map(|ip| ip.into()).collect();
+
+        self
+    }
+
+    /// Consume the builder and register the service
+    pub fn finalise(self) -> Result<()> {
+        let Self {
+            mdns,
+            ips,
+            hostname,
+            ty,
+            path,
+            port,
+            name,
+        } = self;
+
+        let service_type = ty.to_service_type();
+        let domain = format!("{service_type}._tcp.local.");
+        let mut props = HashMap::new();
+
+        props.insert("td".to_string(), path);
+        props.insert("type".to_string(), ty.to_dns_type().to_string());
+
+        let service = ServiceInfo::new(
+            &domain,
+            name.as_ref(),
+            &hostname,
+            ips.as_slice(),
+            port,
+            Some(props),
+        )?;
+
+        mdns.register(service)?;
+
+        Ok(())
+    }
+}
+
+impl Advertiser {
+    /// Create a new service advertiser
+    pub fn new() -> Result<Self> {
+        let mdns = ServiceDaemon::new()?;
+
+        let mut hostname = hostname::get()?.to_string_lossy().to_string();
+        if !hostname.ends_with(".local") {
+            hostname.push_str(".local");
+        }
+        let ips = if_addrs::get_if_addrs()?
+            .iter()
+            .filter_map(|iface| {
+                let ip = iface.ip();
+                match ip {
+                    IpAddr::V4(ip) if !iface.is_loopback() => Some(ip),
+                    _ => None,
+                }
+            })
+            .collect();
+
+        let sa = Self {
+            mdns,
+            ips,
+            hostname,
+        };
+
+        Ok(sa)
+    }
+
+    /// Register and Advertise a new service
+    pub fn add_service(&self, name: impl AsRef<str>) -> ServiceBuilder {
```

Comment:
As a consequence of my previous comment: it could be useful to just take an `impl Into<String>`